### ISSUE TITLE
Reproducing BPD from the paper

### ISSUE THREAD

**Issue by bshall** [pclucas14/vq-vae#1]
Hi @pclucas14,

This is the Benji from Rwanda (not Rosman). It's a nice coincidence that we're both looking into VQVAEs! 

I've been trying to reproduce the results from "Continuous Relaxation Training of Discrete Latent Variable Image Models" and thought you might be interested. I've been able to get the bpd reported in the paper ~4.82 using categorical outputs and 8 codebooks (see [here](https://github.com/bshall/VectorQuantizedVAE)).

From the "Neural Discrete Representation Learning" paper I noticed that they said "we use a field of 32 x 32 latents for ImageNet, or 8 x 8 x 10 for CIFAR10". I think this means that they used 2x downsampling and 10 codebooks for CIFAR10.

**Comment by pclucas14** [comment]
hey Benji! Nice to hear from you :D 

Thanks for sharing your code; I found absolutely no implementations that were reporting BPD and we getting decent results. Do you have any insights to share on training these models  you would like to share ?

I've tried 10 codebooks and 2x downsampling in my code, but was not reaching anything under 5 BPD. 

Best, 
Lucas 

PS: Andre was also working on this and will be pleased to see this (:

**Comment by bshall** [comment]
Nice to hear from you too :)

Yeah, and the paper isn't very detailed about the network at all. As far as I can tell there are 2 main differences between our repos:

1. I used a categorical distribution for the output. Did you experiment with that at all? It takes much longer per epoch so it'd be nice if the discretized logistic loss got the same results.
2. I used batchnorm everywhere and initialized the codebooks to small values.

I haven't done much with on CIFAR10 but in my experiments on speech I couldn't get good codebook usage without using batchnorm. Have you experienced that at all? The idea behind using batchnorm is to make sure that at initialization the continuous encodings are and codes are in the same region.

Lastly, I also vectorized the quantization step for the multiple codebooks in hopes of speeding up training a bit but it didn't help very much.
